File: app/src/main/java/com/chuongvd/app/signal/widget/adapter/BaseRecyclerViewAdapterBinding.java
package com.chuongvd.app.signal.widget.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import com.chuongvd.app.signal.R;
import com.chuongvd.app.signal.databinding.ItemNoDataBinding;
import java.util.ArrayList;
import java.util.List;

public abstract class BaseRecyclerViewAdapterBinding<VIEWHOLDER extends BaseViewHolderBinding, MODEL>
        extends RecyclerView.Adapter<BaseViewHolderBinding> {

    public static final int TYPE_NO_DATA = 0;
    public static final int TYPE_NORMAL = 1;
    public String EMPTY_STRING = "";
    private Context mContext;
    public List<MODEL> mList;
    private LayoutInflater mLayoutInflater;
    protected OnRecyclerItemListener<MODEL> mItemListener;
    private boolean isFirst = true;
    private boolean enableShowNoData = false;

    protected abstract VIEWHOLDER getViewHolder(ViewGroup parent, LayoutInflater inflater);

    public BaseRecyclerViewAdapterBinding(Context context, List<MODEL> list) {
        mContext = context;
        mList = list;
    }

    public BaseRecyclerViewAdapterBinding(List<MODEL> list) {
        mList = list;
    }

    public BaseRecyclerViewAdapterBinding(Context context) {
        mContext = context;
        mList = new ArrayList<>();
    }

    public BaseRecyclerViewAdapterBinding(Context context,
            OnRecyclerItemListener<MODEL> itemListener) {
        mContext = context;
        mList = new ArrayList<>();
        this.mItemListener = itemListener;
    }

    public void setEnableShowNoData(boolean enableShowNoData) {
        this.enableShowNoData = enableShowNoData;
    }

    public void add(MODEL MODEL) {
        getData().add(MODEL);
        notifyItemChanged(getData().size() - 1);
        isFirst = false;
    }

    public void add(List<MODEL> data) {
        int size = getData().size();
        getData().addAll(data);
        if (size == 0) {
            notifyDataSetChanged();
        } else {
            notifyItemInserted(size);
        }
        isFirst = false;
    }

    public void addPreviousEnd(List<MODEL> data) {
        if (data.isEmpty()) return;
        if (getData().isEmpty()) {
            getData().addAll(data);
            notifyDataSetChanged();
        } else {
            int size = getData().size();
            getData().addAll(data);
            notifyItemRangeInserted(size - 1, data.size());
        }
        isFirst = false;
    }

    public void setData(List<MODEL> data) {
        getData().clear();
        getData().addAll(data);
        notifyDataSetChanged();
        isFirst = false;
    }

    public synchronized void remove(MODEL MODEL) {
        int position = -1;
        for (int i = 0; i < getData().size(); i++) {
            if (!MODEL.equals(getData().get(i))) {
                continue;
            }
            position = i;
            break;
        }

        if (position == -1) return;
        removePosition(position);
        isFirst = false;
    }

    public synchronized void removePosition(int position) {
        if (getData().size() < position) return;
        getData().remove(position);
        notifyItemChanged(position);
        isFirst = false;
    }

    public boolean isEmpty() {
        return mList.size() == 0;
    }

    public void setItemListener(OnRecyclerItemListener<MODEL> listener) {
        mItemListener = listener;
        notifyDataSetChanged();
    }

    public synchronized void remove(List<MODEL> data) {
        getData().removeAll(data);
        notifyDataSetChanged();
        isFirst = false;
    }

    public synchronized void removeAll() {
        getData().clear();
        notifyDataSetChanged();
        isFirst = false;
    }

    public synchronized void refreshData(List<MODEL> data) {
        getData().clear();
        getData().addAll(data);
        notifyDataSetChanged();
        isFirst = false;
    }

    public void setFirst(boolean first) {
        isFirst = first;
    }

    public List<MODEL> getData() {
        return mList == null ? mList = new ArrayList<>() : mList;
    }

    private LayoutInflater getLayoutInflater(Context context) {
        mContext = context;
        return mLayoutInflater == null ? mLayoutInflater = LayoutInflater.from(context)
                : mLayoutInflater;
    }

    public Context getContext() {
        return mContext;
    }

    @Override
    public BaseViewHolderBinding onCreateViewHolder(ViewGroup parent, int viewType) {
        BaseViewHolderBinding baseViewHolderBinding;
        if (viewType == TYPE_NO_DATA) {
            baseViewHolderBinding =
                    getNoDataViewHolder(parent, getLayoutInflater(parent.getContext()));
        } else {
            baseViewHolderBinding = getViewHolder(parent, getLayoutInflater(parent.getContext()));
        }
        return baseViewHolderBinding;
    }

    public BaseViewHolderBinding getNoDataViewHolder(ViewGroup parent, LayoutInflater inflater) {
        return new NoDataViewHolder(ItemNoDataBinding.inflate(inflater, parent, false));
    }

    public void setEmptyString(String emptyString) {
        EMPTY_STRING = emptyString;
        notifyDataSetChanged();
    }

    @Override
    public void onBindViewHolder(BaseViewHolderBinding holder, int position) {
        if (getData().size() > 0) {
            holder.bindData(mList.get(position));
        } else {
            if (!isFirst && enableShowNoData) {
                holder.bindData(EMPTY_STRING);
            }
            if (isFirst) {
                isFirst = false;
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (getData().size() == 0) {
            return TYPE_NO_DATA;
        }
        return TYPE_NORMAL;
    }

    @Override
    public int getItemCount() {
        if (getData().size() == 0) {
            return 1;
        }
        return getData().size();
    }

    public void softUpdateListData(List<MODEL> newData) {
        if (newData == null || newData.isEmpty()) {
            mList.clear();
            return;
        }
        int newSize = newData.size();
        int oldSize = mList.size();
        int minSize = Math.min(oldSize, newSize);
        int maxSize = Math.max(oldSize, newSize);
        for (int i = 0; i < minSize; i++) {
            MODEL data = newData.get(i);
            if (!data.equals(mList.get(i))) {
                mList.remove(i);
                mList.add(i, data);
                notifyItemChanged(i);
            }
        }
        for (int i = minSize; i < maxSize; i++) {
            if (newSize > oldSize) {
                mList.add(newData.get(i));
                notifyItemRemoved(i);
            } else {
                mList.remove(newSize);
                notifyItemRemoved(newSize);
            }
        }
    }

    class NoDataViewHolder extends BaseViewHolderBinding<ItemNoDataBinding, String> {

        NoDataViewHolder(ItemNoDataBinding binding) {
            super(binding);
        }

        @Override
        public void bindData(String s) {
            if (TextUtils.isEmpty(s)) {
                s = getContext().getString(R.string.no_data);
            }
            mBinding.setText(s);
        }
    }

    public interface OnRecyclerItemListener<MODEL> {
        void onItemClick(int position, MODEL data);
    }
}

File: app/src/main/java/com/chuongvd/app/signal/widget/adapter/BaseRecyclerViewAdapterBinding.kt
package com.chuongvd.app.signal.widget.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.chuongvd.app.signal.databinding.ItemNoDataBinding
import java.util.*

abstract class BaseRecyclerViewAdapterBinding<VIEWHOLDER : BaseViewHolderBinding<*, MODEL>, MODEL> : RecyclerView.Adapter<BaseViewHolderBinding<*, MODEL>> {
  var EMPTY_STRING = ""
  var context: Context? = null
    private set
  var mList: MutableList<MODEL> = mutableListOf()
  private var mLayoutInflater: LayoutInflater? = null
  protected var mItemListener: OnRecyclerItemListener<MODEL>? = null
  private var isFirst = true
  private var enableShowNoData = false

  val isEmpty: Boolean
    get() = mList.size == 0

  var data: MutableList<MODEL>
    get() = mList
    set(data) {
      data.clear()
      data.addAll(data)
      notifyDataSetChanged()
      isFirst = false
    }

  protected abstract fun getViewHolder(parent: ViewGroup, inflater: LayoutInflater): VIEWHOLDER

  constructor(context: Context, list: MutableList<MODEL>) {
    this.context = context
    mList = list
  }

  constructor(list: MutableList<MODEL>) {
    mList = list
  }

  constructor(context: Context) {
    this.context = context
    mList = ArrayList()
  }

  constructor(context: Context,
      itemListener: OnRecyclerItemListener<MODEL>) {
    this.context = context
    mList = ArrayList()
    this.mItemListener = itemListener
  }

  fun setEnableShowNoData(enableShowNoData: Boolean) {
    this.enableShowNoData = enableShowNoData
  }

  fun add(MODEL: MODEL) {
    data.add(MODEL)
    notifyItemChanged(data.size - 1)
    isFirst = false
  }

  fun add(list: List<MODEL>) {
    val size = this.data.size
    this.data.addAll(list)
    if (size == 0) {
      notifyDataSetChanged()
    } else {
      notifyItemInserted(size)
    }
    isFirst = false
  }

  fun addPreviousEnd(list: List<MODEL>) {
    if (list.isEmpty()) return
    if (this.data.isEmpty()) {
      this.data.addAll(list)
      notifyDataSetChanged()
    } else {
      val size = this.data.size
      data.addAll(list)
      notifyItemRangeInserted(size - 1, this.data.size)
    }
    isFirst = false
  }

  @Synchronized
  fun remove(MODEL: MODEL) {
    var position = -1
    for (i in 0 until data.size) {
      if (MODEL != data[i]) {
        continue
      }
      position = i
      break
    }

    if (position == -1) return
    removePosition(position)
    isFirst = false
  }

  @Synchronized
  fun removePosition(position: Int) {
    if (data.size < position) return
    data.removeAt(position)
    notifyItemChanged(position)
    isFirst = false
  }

  fun setItemListener(listener: OnRecyclerItemListener<MODEL>) {
    mItemListener = listener
    notifyDataSetChanged()
  }

  @Synchronized
  fun remove(list: List<MODEL>) {
    data.removeAll(list)
    notifyDataSetChanged()
    isFirst = false
  }

  @Synchronized
  fun removeAll() {
    data.clear()
    notifyDataSetChanged()
    isFirst = false
  }

  @Synchronized
  fun refreshData(list: List<MODEL>) {
    data.clear()
    data.addAll(list)
    notifyDataSetChanged()
    isFirst = false
  }

  fun setFirst(first: Boolean) {
    isFirst = first
  }

  private fun getLayoutInflater(context: Context): LayoutInflater {
    this.context = context
    if (mLayoutInflater == null)
      mLayoutInflater = LayoutInflater.from(context)
    else
      mLayoutInflater
    return mLayoutInflater!!
  }

  override fun onCreateViewHolder(parent: ViewGroup,
      viewType: Int): BaseViewHolderBinding<*, MODEL> {
    val baseViewHolderBinding: BaseViewHolderBinding<*, MODEL>
    if (viewType == TYPE_NO_DATA) {
      baseViewHolderBinding = getNoDataViewHolder(parent,
          getLayoutInflater(parent.context)) as BaseViewHolderBinding<*, MODEL>
    } else {
      baseViewHolderBinding = getViewHolder(parent, getLayoutInflater(parent.context))
    }
    return baseViewHolderBinding
  }

  fun getNoDataViewHolder(parent: ViewGroup,
      inflater: LayoutInflater): NoDataViewHolder {
    return NoDataViewHolder(context!!, ItemNoDataBinding.inflate(inflater, parent, false))
  }

  fun setEmptyString(emptyString: String) {
    EMPTY_STRING = emptyString
    notifyDataSetChanged()
  }

  override fun onBindViewHolder(holder: BaseViewHolderBinding<*, MODEL>, position: Int) {
    if (data.size > 0) {
      holder.bindData(data.get(position))
    } else {
      if (!isFirst && enableShowNoData) {
        holder.bindData(EMPTY_STRING as MODEL)
      }
      if (isFirst) {
        isFirst = false
      }
    }
  }

  override fun getItemViewType(position: Int): Int {
    return if (data.size == 0) {
      TYPE_NO_DATA
    } else TYPE_NORMAL
  }

  override fun getItemCount(): Int {
    return if (data.size == 0) {
      1
    } else data.size
  }

  fun softUpdateListData(newData: List<MODEL>?) {
    if (newData == null || newData.isEmpty()) {
      mList.clear()
      return
    }
    val newSize = newData.size
    val oldSize = mList.size
    val minSize = Math.min(oldSize, newSize)
    val maxSize = Math.max(oldSize, newSize)
    for (i in 0 until minSize) {
      val data = newData[i]
      if (data != mList[i]) {
        mList.removeAt(i)
        mList.add(i, data)
        notifyItemChanged(i)
      }
    }
    for (i in minSize until maxSize) {
      if (newSize > oldSize) {
        mList.add(newData[i])
        notifyItemRemoved(i)
      } else {
        mList.removeAt(newSize)
        notifyItemRemoved(newSize)
      }
    }
  }

  interface OnRecyclerItemListener<MODEL> {
    fun onItemClick(position: Int, data: MODEL)
  }

  companion object {

    val TYPE_NO_DATA = 0
    val TYPE_NORMAL = 1
  }
}
